### PR TITLE
Make it possible to run Load and Density tests using Deployments or ReplicaSets

### PR DESCRIPTION
**Pull request by gmarek**
This is a first extension to our scalability tests.

cc @timothysc @jeremyeder

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37696)

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -2757,47 +2757,117 @@ func WaitForPodsWithLabel(c clientset.Interface, ns string, label labels.Selecto
...
+	case *extensionsinternal.ReplicaSet:
+		return typed.Spec.Replicas, nil
+	case *extensionsinternal.Deployment:
+		return typed.Spec.Replicas, nil
+	default:
+		return -1, fmt.Errorf("Unsupported kind: %v", obj)
+	}
+}
+
+func getReaperForKind(clientset clientset.Interface, internalClientset internalclientset.Interface, kind schema.GroupKind) (kubectl.Reaper, error) {
```
Comment (wojtek-t) [inline test/e2e/framework/util.go]: You don't need "clientset" in this function.

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -2757,47 +2757,117 @@ func WaitForPodsWithLabel(c clientset.Interface, ns string, label labels.Selecto
...
+	case extensionsinternal.Kind("ReplicaSet"):
+		return c.Extensions().ReplicaSets(ns).Delete(name, deleteOption)
+	case extensionsinternal.Kind("Deployment"):
+		return c.Extensions().Deployments(ns).Delete(name, deleteOption)
+	default:
+		return fmt.Errorf("Unsupported kind: %v", kind)
+	}
+}
+
+func getSelectorFromRuntimeObject(obj runtime.Object) (*unversioned.LabelSelector, error) {
```
Comment (wojtek-t) [inline test/e2e/framework/util.go]: We should return "labels.Selector" from this method (as this is the generic interface). The further code will be simpler then.

File: test/e2e/density.go [hunk 3]
```diff
@@ -199,7 +199,7 @@ func runDensityTest(dtc DensityTestConfig) time.Duration {
 			// Call wg.Done() in defer to avoid blocking whole test
 			// in case of error from RunRC.
 			defer wg.Done()
-			framework.ExpectNoError(framework.RunRC(rcConfig))
+			framework.ExpectNoError(rcConfig.Run())
```
Comment (wojtek-t) [inline test/e2e/density.go]: We should rename rcConfig to objConfig or sth like that.

File: test/e2e/framework/util.go [hunk 4]
```diff
@@ -2803,16 +2803,22 @@ func getReplicasFromRuntimeObject(obj runtime.Object) (int32, error) {
...
+	case *extensions.ReplicaSet:
+		if typed.Spec.Replicas != nil {
+			return *typed.Spec.Replicas, nil
+		}
+		return 0, nil
+	case *extensions.Deployment:
+		if typed.Spec.Replicas != nil {
+			return *typed.Spec.Replicas, nil
+		}
+		return 0, nil
```
Comment (wojtek-t) [inline test/e2e/framework/util.go]: The first two commits are touching the same code - can you please squash those two?

File: test/utils/runners.go [hunk 5]
```diff
@@ -193,6 +202,22 @@ func RunDeployment(config DeploymentConfig) error {
 	return config.start()
 }
 
+func (config *DeploymentConfig) Run() error {
+	return RunDeployment(*config)
+}
+
+func (config *DeploymentConfig) GetName() string {
```
Comment (wojtek-t) [inline test/utils/runners.go]: I think that you don't need to redefine GetName() and GetNamespace() if you are just calling the same method from base object.

File: test/e2e/framework/util.go [hunk 6]
```diff
@@ -2659,51 +2679,37 @@ func ScaleRC(clientset clientset.Interface, internalClientset internalclientset.
...
+	return WaitForControlledPodsRunning(clientset, ns, name, kind)
 }
 
 // Wait up to 10 minutes for pods to become Running.
-func WaitForRCPodsRunning(c clientset.Interface, ns, rcName string) error {
-	rc, err := c.Core().ReplicationControllers(ns).Get(rcName)
+func WaitForControlledPodsRunning(c clientset.Interface, ns, name string, kind schema.GroupKind) error {
+	rtObject, err := getRuntimeObjectForKind(c, kind, ns, name)
 	if err != nil {
+		if apierrs.IsNotFound(err) {
```
Comment (wojtek-t) [inline test/e2e/framework/util.go]: This means we can't wait for pods, so in my opinion not having this object is an error (even if it doesn't exist). I would remove this if.

File: test/e2e/density.go [hunk 7]
```diff
@@ -361,18 +367,29 @@ var _ = framework.KubeDescribe("Density", func() {
 		{podsPerNode: 50, runLatencyTest: false},
 		{podsPerNode: 95, runLatencyTest: true},
 		{podsPerNode: 100, runLatencyTest: false},
+		{podsPerNode: 30, runLatencyTest: true, kind: extensions.Kind("Deployment")},
```
Comment (wojtek-t) [inline test/e2e/density.go]: I wouldn't introduce those as predefined. I think that we should just mix all types (but not in this PR).

File: test/e2e/load.go [hunk 8]
```diff
@@ -117,20 +119,24 @@ var _ = framework.KubeDescribe("Load capacity", func() {
...
+		// What kind of resource we want to create
+		kind schema.GroupKind
 	}
 
 	loadTests := []Load{
 		// The container will consume 1 cpu and 512mb of memory.
-		{podsPerNode: 3, image: "jess/stress", command: []string{"stress", "-c", "1", "-m", "2"}},
-		{podsPerNode: 30, image: "gcr.io/google_containers/serve_hostname:v1.4"},
+		{podsPerNode: 3, image: "jess/stress", command: []string{"stress", "-c", "1", "-m", "2"}, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 30, image: "gcr.io/google_containers/serve_hostname:v1.4", kind: api.Kind("ReplicationController")},
```
Comment (wojtek-t) [inline test/e2e/load.go]: I wouldn't change predefined tests. We should have a test that has everything inside, not 3 different tests. [Obviously not in this PR].

File: test/utils/runners.go [hunk 9]
```diff
@@ -218,6 +223,26 @@ func (config *DeploymentConfig) GetKind() schema.GroupKind {
 	return extensionsinternal.Kind("Deployment")
 }
 
+func (config *DeploymentConfig) GetClient() clientset.Interface {
```
Comment (wojtek-t) [inline test/utils/runners.go]: Again - if you are using method from underlying object, I think you don't need to redefine it.

File: test/e2e/framework/util.go (hunk 2 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Done.

File: test/e2e/framework/util.go (hunk 1 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Done in one of later commits.

File: test/e2e/density.go (hunk 3 above)
Comment (gmarek) [inline test/e2e/density.go]: Done in one of later commits.

File: test/e2e/framework/util.go (hunk 4 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Done.

File: test/e2e/framework/util.go (hunk 6 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Done.

File: test/e2e/density.go (hunk 7 above)
Comment (gmarek) [inline test/e2e/density.go]: I think it's useful to have them here - so we'll be able to easily run them to diagnose possible problems.

File: test/utils/runners.go (hunk 9 above)
Comment (gmarek) [inline test/utils/runners.go]: Done.

File: test/utils/runners.go (hunk 5 above)
Comment (gmarek) [inline test/utils/runners.go]: Done.

File: test/e2e/density.go (hunk 7 above)
Comment (wojtek-t) [inline test/e2e/density.go]: OK. But let's treat this PR as purely infrastructural, and let's add this in a separate PR.

File: test/e2e/density.go [hunk 10]
```diff
@@ -423,6 +439,17 @@ var _ = framework.KubeDescribe("Density", func() {
 					MaxContainerFailures: &MaxContainerFailures,
 					Silent:               true,
 				}
+				switch itArg.kind {
+				case api.Kind("ReplicationController"):
+					configs[i] = baseConfig
+				case extensions.Kind("ReplicaSet"):
+					framework.Logf("Creating ReplicaSet")
```
Comment (wojtek-t) [inline test/e2e/density.go]: Please unify logging

File: test/e2e/load.go [hunk 11]
```diff
@@ -46,9 +48,9 @@ const (
 	smallRCSize       = 5
 	mediumRCSize      = 30
 	bigRCSize         = 250
-	smallRCGroupName  = "load-small-rc"
-	mediumRCGroupName = "load-medium-rc"
-	bigRCGroupName    = "load-big-rc"
+	smallRCGroupName  = "load-small"
```
Comment (wojtek-t) [inline test/e2e/load.go]: If you're touching it, you should probably rename it to "smallGroupName" (without RC)? Or sth like that.

File: test/e2e/load.go (hunk 8 above)
Comment (wojtek-t) [inline test/e2e/load.go]: Same  here - let's treat this PR as infrastructural and move this one to a separate PR if you want this.

File: test/e2e/load.go [hunk 12]
```diff
@@ -311,20 +317,32 @@ func generateRCConfigsForGroup(
 			CpuRequest:     10,       // 0.01 core
 			MemRequest:     26214400, // 25MB
 		}
+		var config testutils.RunObjectConfig
+		switch kind {
+		case api.Kind("ReplicationController"):
+			config = baseConfig
+		case extensions.Kind("ReplicaSet"):
+			framework.Logf("Creating ReplicaSet")
```
Comment (wojtek-t) [inline test/e2e/load.go]: Please unify lgging.

File: test/e2e/load.go [hunk 13]
```diff
@@ -343,69 +361,67 @@ func sleepUpTo(d time.Duration) {
...
-	options := v1.ListOptions{
-		LabelSelector:   selector.String(),
-		ResourceVersion: "0",
-	}
-	_, err := config.Client.Core().Pods(config.Namespace).List(options)
-	framework.ExpectNoError(err, fmt.Sprintf("listing pods from rc %v", config.Name))
+	newSize := uint(rand.Intn(config.GetReplicas()) + config.GetReplicas()/2)
+	framework.ExpectNoError(framework.ScaleResource(
+		config.GetClient(), config.GetInternalClient(), config.GetNamespace(), config.GetName(), newSize, true, config.GetKind()),
+		fmt.Sprintf("scaling rc %s for the first time", config.GetName()))
```
Comment (wojtek-t) [inline test/e2e/load.go]: Please revert the change with List - this was added explicitly to test listing too.

File: test/utils/runners.go [hunk 14]
```diff
@@ -338,6 +343,26 @@ func (config *RCConfig) GetKind() schema.GroupKind {
...
+}
+
+func (config *RCConfig) SetInternalClient(c internalclientset.Interface) {
+	config.InternalClient = c
+}
+
+func (config *RCConfig) GetReplicas() int {
+	return config.Replicas
+}
+
```
Comment (wojtek-t) [inline test/utils/runners.go]: Since all commits are actually touching the same code, can you please squash all of them?

File: test/e2e/load.go (hunk 8 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.

File: test/e2e/density.go (hunk 10 above)
Comment (gmarek) [inline test/e2e/density.go]: Done.

File: test/e2e/load.go (hunk 11 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.

File: test/e2e/load.go (hunk 12 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.

File: test/e2e/load.go (hunk 13 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.

File: test/utils/runners.go (hunk 14 above)
Comment (gmarek) [inline test/utils/runners.go]: Done.

**Comment by jeremyeder** [comment]
@sjug please review and comment.

File: test/e2e/density.go [hunk 15]
```diff
@@ -352,27 +356,38 @@ var _ = framework.KubeDescribe("Density", func() {
...
+		{podsPerNode: 95, runLatencyTest: true, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 100, runLatencyTest: false, kind: api.Kind("ReplicationController")},
 	}
 
 	for _, testArg := range densityTests {
 		feature := "ManualPerformance"
+		if testArg.kind.Empty() {
+			testArg.kind = api.Kind("ReplicationController")
+		}
+		if testArg.kind != extensions.Kind("Deployment") && testArg.kind != extensions.Kind("ReplicaSet") && testArg.kind != api.Kind("ReplicationController") {
```
Comment (wojtek-t) [inline test/e2e/density.go]: Let's remove this if - it shouldn't be needed to check if test is written correctly :)

File: test/e2e/density.go [hunk 16]
```diff
@@ -352,27 +356,38 @@ var _ = framework.KubeDescribe("Density", func() {
...
+		{podsPerNode: 3, runLatencyTest: false, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 30, runLatencyTest: true, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 50, runLatencyTest: false, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 95, runLatencyTest: true, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 100, runLatencyTest: false, kind: api.Kind("ReplicationController")},
 	}
 
 	for _, testArg := range densityTests {
 		feature := "ManualPerformance"
+		if testArg.kind.Empty() {
```
Comment (wojtek-t) [inline test/e2e/density.go]: Can we get rid of this if and assume that this is set?

File: test/e2e/load.go [hunk 17]
```diff
@@ -117,20 +120,28 @@ var _ = framework.KubeDescribe("Load capacity", func() {
...
-		{podsPerNode: 3, image: "jess/stress", command: []string{"stress", "-c", "1", "-m", "2"}},
-		{podsPerNode: 30, image: "gcr.io/google_containers/serve_hostname:v1.4"},
+		{podsPerNode: 3, image: "jess/stress", command: []string{"stress", "-c", "1", "-m", "2"}, kind: api.Kind("ReplicationController")},
+		{podsPerNode: 30, image: "gcr.io/google_containers/serve_hostname:v1.4", kind: api.Kind("ReplicationController")},
 	}
 
 	for _, testArg := range loadTests {
 		feature := "ManualPerformance"
-		if testArg.podsPerNode == 30 {
+		if testArg.kind.Empty() {
```
Comment (wojtek-t) [inline test/e2e/load.go]: Same here - can we assume it's set and remove this if?

File: test/e2e/load.go [hunk 18]
```diff
@@ -117,20 +120,28 @@ var _ = framework.KubeDescribe("Load capacity", func() {
...
+			testArg.kind = api.Kind("ReplicationController")
+		}
+		if testArg.kind != extensions.Kind("Deployment") && testArg.kind != extensions.Kind("ReplicaSet") && testArg.kind != api.Kind("ReplicationController") {
```
Comment (wojtek-t) [inline test/e2e/load.go]: Let's remove this if - it shouldn't be needed to check if test is written correctly :)

**Review by wojtek-t — COMMENTED** [review]
4 very minor comments - other than that lgtm

File: test/e2e/density.go (hunk 16 above)
Comment (gmarek) [inline test/e2e/density.go]: Done.

File: test/e2e/density.go (hunk 15 above)
Comment (gmarek) [inline test/e2e/density.go]: Done.

File: test/e2e/load.go (hunk 17 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.

File: test/e2e/load.go (hunk 18 above)
Comment (gmarek) [inline test/e2e/load.go]: Done.
